client/beefy/src/worker.rs

```diff
@@ -351,55 +349,171 @@ where
+			hash
+		} else {
+			warn!(target: "beefy", "🥩 No MMR root digest found for: {:?}", target_hash);
+			return
+		};
+		let payload = Payload::new(known_payload_ids::MMR_ROOT_ID, mmr_root.encode());
+
+		let (validators, validator_set_id) = if let Some(rounds) = &self.rounds {
+			if !rounds.should_self_vote(&(payload.clone(), target_number)) {
+				debug!(target: "beefy", "🥩 Don't double vote for block number: {:?}", target_number);
+				return
+			}
+			(rounds.validators_for(target_number), rounds.validator_set_id_for(target_number))
+		} else {
+			debug!(target: "beefy", "🥩 Missing validator set - can't vote for: {:?}", target_hash);
+			return
+		};
+		let authority_id = if let Some(id) = self.key_store.authority_id(validators) {
+			debug!(target: "beefy", "🥩 Local authority id: {:?}", id);
+			id
+		} else {
+			debug!(target: "beefy", "🥩 Missing validator id - can't vote for: {:?}", target_hash);
+			return
+		};
+
+		let commitment = Commitment { payload, block_number: target_number, validator_set_id };
+		let encoded_commitment = commitment.encode();
+
+		let signature = match self.key_store.sign(&authority_id, &*encoded_commitment) {
+			Ok(sig) => sig,
+			Err(err) => {
+				warn!(target: "beefy", "🥩 Error signing commitment: {:?}", err);
+				return
+			},
+		};
+
+		trace!(
+			target: "beefy",
+			"🥩 Produced signature using {:?}, is_valid: {:?}",
+			authority_id,
+			BeefyKeystore::verify(&authority_id, &signature, &*encoded_commitment)
+		);
+
+		let message = VoteMessage { commitment, id: authority_id, signature };
+
+		let encoded_message = message.encode();
 
-				metric_set!(self, beefy_best_block, block_num);
+		metric_inc!(self, beefy_votes_sent);
+
+		debug!(target: "beefy", "🥩 Sent vote message: {:?}", message);
+
+		self.handle_vote(
+			(message.commitment.payload, message.commitment.block_number),
+			(message.id, message.signature),
+			true,
+		);
+
+		self.gossip_engine.lock().gossip_message(topic::<B>(), encoded_message, false);
+	}
+
+	/// Wait for BEEFY runtime pallet to be available.
+	#[cfg(not(test))]
+	async fn wait_for_runtime_pallet(&mut self) {
+		loop {
+			let at = BlockId::hash(self.best_grandpa_block_header.hash());
+			if let Some(active) = self.client.runtime_api().validator_set(&at).ok().flatten() {
+				if active.id() == GENESIS_AUTHORITY_SET_ID {
+					// When starting from genesis, there is no session boundary digest.
+					// Just initialize `rounds` to Block #1 as BEEFY mandatory block.
+					self.init_session_at(active, 1u32.into());
+				}
+				// In all other cases, we just go without `rounds` initialized, meaning the worker
+				// won't vote until it witnesses a session change.
+				// Once we'll implement 'initial sync' (catch-up), the worker will be able to start
+				// voting right away.
+				break
+			} else {
+				info!(target: "beefy", "Waiting BEEFY pallet to become available...");
+				sleep(Duration::from_secs(5)).await;
 			}
 		}
 	}
```

Comment:
I don't think this will work as `at` will never change throughout the loop. Can we instead listen to finality notifications until the runtime call works? 

```rust
block_finality_notifications().take_while(|notif| {
   runtime_api().validator_set(notif.block)
   // ...
})
```